### PR TITLE
Generalize the Dictionary to hold arbitrary columns

### LINKED ISSUE #99: Generalize the Dictionary to hold arbitrary columns
Don't hardcode the column names in the `Dictionary`. Make it possible to store arbitrary columns (string, date and pango_lineage).

### PR DESCRIPTION
**Pull request by JonasKellerer**
Resolves #99 

Needs #100 and is followed by #101 to fully take effect.

### CODE REVIEW

File: src/silo/storage/dictionary.cpp [hunk 1]
```diff
@@ -12,6 +12,45 @@
...
+
+   for (const auto& column_name : pango_lineage_column_names) {
+      pango_lineage_lookup[column_name] = std::unordered_map<PangoLineage, ValueId>();
+      reverse_pango_lineage_lookup[column_name] = std::vector<PangoLineage>();
+   }
+}
+
+Dictionary::Dictionary()
+    : Dictionary(
+         std::vector<std::string>({"country", "region", "division", "column"}),
```
Comment (fengelniederhammer) [inline src/silo/storage/dictionary.cpp:50]: This looks like a leftover?
```suggestion
         std::vector<std::string>({"country", "region", "division"}),
```

File: src/silo/storage/dictionary.cpp [hunk 2]
```diff
@@ -56,302 +95,110 @@ void Dictionary::updateDictionary(
...
 uint32_t Dictionary::getPangoLineageIdInLookup(const std::string& pango_lineage) const {
-   if (pango_lineage_dictionary.contains(pango_lineage)) {
-      return pango_lineage_dictionary.at(pango_lineage);
+   if (pango_lineage_lookup.at("pango_lineage").contains(pango_lineage)) {
+      return pango_lineage_lookup.at("pango_lineage").at(pango_lineage);
    }
    return UINT32_MAX;
 }
 
 std::string error_string = "NOID";
```
Comment (fengelniederhammer) [inline src/silo/storage/dictionary.cpp:126]: According to my IDE, this is unused. Can we delete it?

File: src/silo/storage/dictionary.cpp [hunk 3]
```diff
@@ -56,302 +95,110 @@ void Dictionary::updateDictionary(
...
-   return UINT32_MAX;
+   return std::nullopt;
 }
 
-[[maybe_unused]] const std::string& Dictionary::getColumn(uint32_t column_id_in_lookup) const {
-   if (column_id_in_lookup == UINT32_MAX) {
-      return error_string;
+std::optional<std::string> Dictionary::lookupPangoLineageValue(
+   const std::string& column_name,
+   uint32_t value_id
```
Comment (fengelniederhammer) [inline src/silo/storage/dictionary.cpp:196]:
```suggestion
   ValueId value_id
```

File: src/silo/storage/dictionary.cpp [hunk 4]
```diff
@@ -56,302 +95,110 @@ void Dictionary::updateDictionary(
...
+   return UINT32_MAX;
+}
+// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+std::string Dictionary::getColumn(uint32_t /*column_id_in_lookup*/) const {
+   return "";
 }
 
-[[maybe_unused]] const std::string& Dictionary::getGeneralLookup(uint64_t general_id_in_lookup
+std::optional<ValueId> Dictionary::lookupValueId(
+   const std::string& column_name,
```
Comment (fengelniederhammer) [inline src/silo/storage/dictionary.cpp:175]: Let's also use the alias here
```suggestion
   const ColumnName& column_name,
```

File: include/silo/storage/dictionary.h [hunk 5]
```diff
@@ -3,35 +3,38 @@
...
-   uint32_t region_count = 0;
-   uint32_t additional_columns_count = 0;
-   uint64_t general_count = 0;
+   std::unordered_map<ColumnName, std::unordered_map<std::string, ValueId>> string_lookup;
+   std::unordered_map<ColumnName, std::vector<std::string>> reverse_string_lookup;
+   void updateStringLookup(const ColumnName& column_name, const std::string& value);
+
+   std::unordered_map<ColumnName, std::unordered_map<PangoLineage, ValueId>> pango_lineage_lookup;
+   std::unordered_map<ColumnName, std::vector<PangoLineage>> reverse_pango_lineage_lookup;
+   void updatePangoLineageLookup(const ColumnName& column_name, const PangoLineage& pango_lineage);
```
Comment (fengelniederhammer) [inline include/silo/storage/dictionary.h:29]: We could probably also encapsulate this and deduplicate it? Does that make sense? And maybe find a better name than "TypedLookup".
```suggestion
   TypedLookup<std::string> string_lookup;
   TypedLookup<PangoLineage> pango_lineage_lookup;
```

File: src/silo/storage/dictionary.test.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+   ASSERT_FALSE(dictionary.lookupValueId("country", "NotACountry").has_value());
+}
+
+TEST_F(DictionaryTestFixture, shouldLookupValueId) {
+   ASSERT_EQ(dictionary.lookupValueId("country", "Switzerland"), 0);
+   ASSERT_EQ(dictionary.lookupValueId("division", "Bern"), 1);
+
+   ASSERT_EQ(
+      dictionary.lookupValueId("division", "Bern"), dictionary.lookupValueId("division", "Bern")
+   );
```
Comment (fengelniederhammer) [inline src/silo/storage/dictionary.test.cpp:48]: Why do we need that assertion?^^

**Review by fengelniederhammer — APPROVED** [review]
Just a couple of minor things, overall it looks good :)

File: src/silo/storage/dictionary.test.cpp (hunk 6 above)
Comment (JonasKellerer) [reply src/silo/storage/dictionary.test.cpp:48]: I thought that reading a value twice should still result in the same value. But is not neccessary per se :)

File: include/silo/storage/dictionary.h (hunk 5 above)
Comment (JonasKellerer) [reply include/silo/storage/dictionary.h]: Good idea.

**Pull request merged**